src/Chapter3.hs

```diff
@@ -1006,6 +1142,189 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+-- Couldn't make it simpler :(
+data KnightAction
+  = KnightAttack
+  | Drink
+  | Cast
+  deriving (Show)
+
+data MonsterAction
+  = MonsterAttack
+  | Run
+  deriving (Show)
+
+data Knight = Knight
+  { knightHealth :: Int
+  , knightAttack :: Int
+  , knightDefense :: Int
+  , knightActions :: [KnightAction]
+  , baseKnightHealth :: Int
+  , baseKnightHealthIncrease :: Int
+  , baseKnightDefense :: Int
+  , baseKnightDefenseIncrease :: Int
+  } deriving (Show)
+
+data Monster = Monster
+  { monsterHealth :: Int
+  , monsterAttack :: Int
+  , monsterActions :: [MonsterAction]
+  } deriving (Show)
+
+data ActionType
+  = GenericAttack
+  | SpecialAction
+  deriving (Show)
+
+getFirstKnightAction :: Knight -> KnightAction
+getFirstKnightAction Knight { knightActions = (x:_) } = x
+getFirstKnightAction _ = error "Forgot to 'cycle' knight actions"
+
+getFirstMonsterAction :: Monster -> MonsterAction
+getFirstMonsterAction Monster { monsterActions = (x:_) } = x
+getFirstMonsterAction _ = error "Forgot to 'cycle' monster actions"
+
+getFollowingKnightActions :: Knight -> [KnightAction]
+getFollowingKnightActions Knight { knightActions = (_:xs) } = xs
+getFollowingKnightActions _ = error "Forgot to 'cycle' knight actions"
```

Comment:
You can combine this and `getFirstKnightAction` functions.

One option is to return a tuple:

```haskell
unconsFollowingKnightActions :: Knight -> (KnightAction, [KnightAction])
```

But best of all is to avoid `errors again here.

You can instead return `Maybe`:

```haskell
unconsFollowingKnightActions :: Knight -> Maybe (KnightAction, [KnightAction])
```

Ans also you can reuse `uncons` standard function for that.

Reply:
Using a tuple for this was my first thought, but I couldn't figure out how to use both values after I got them.

Reply:
There are a few options here. 
1. Use `fst` and `snd` functions to take fist or second element of the tuple
2. Pattern match on that giving the name to each element of the pair:
    ```haskell
    where
      (curAction, restActions) = unconsFollowingKnightActions knight
    ```
    and you can use `curAction` and `restAction` as separate elements in your code 🙂 

Hope it makes sense 🙂

Reply:
Yes, it clicks. Thanks!